pkg/cephfs/volume.go

```diff
@@ -46,111 +36,27 @@ func getVolumeNamespace(volID volumeID) string {
-		return err
-	}
-	defer unmountCephRoot(volID)
-
-	var (
-		volRoot         = getCephRootVolumePathLocal(volID)
-		volRootCreating = volRoot + "-creating"
-	)
-
-	if pathExists(volRoot) {
-		klog.V(4).Infof("cephfs: volume %s already exists, skipping creation", volID)
-		return nil
-	}
-
-	if err := createMountPoint(volRootCreating); err != nil {
-		return err
-	}
-
-	if bytesQuota > 0 {
-		if err := setVolumeAttribute(volRootCreating, "ceph.quota.max_bytes", fmt.Sprintf("%d", bytesQuota)); err != nil {
-			return err
-		}
-	}
-
-	if err := setVolumeAttribute(volRootCreating, "ceph.dir.layout.pool", volOptions.Pool); err != nil {
-		return fmt.Errorf("%v\ncephfs: Does pool '%s' exist?", err, volOptions.Pool)
-	}
-
-	if err := setVolumeAttribute(volRootCreating, "ceph.dir.layout.pool_namespace", getVolumeNamespace(volID)); err != nil {
-		return err
-	}
-
-	if err := os.Rename(volRootCreating, volRoot); err != nil {
-		return fmt.Errorf("couldn't mark volume %s as created: %v", volID, err)
-	}
-
-	return nil
-}
-
-func purgeVolume(volID volumeID, adminCr *credentials, volOptions *volumeOptions) error {
-	if err := mountCephRoot(volID, volOptions, adminCr); err != nil {
-		return err
-	}
-	defer unmountCephRoot(volID)
-
-	var (
-		volRoot         = getCephRootVolumePathLocal(volID)
-		volRootDeleting = volRoot + "-deleting"
-	)
-
-	if pathExists(volRoot) {
-		if err := os.Rename(volRoot, volRootDeleting); err != nil {
-			return fmt.Errorf("couldn't mark volume %s for deletion: %v", volID, err)
-		}
-	} else {
-		if !pathExists(volRootDeleting) {
-			klog.V(4).Infof("cephfs: volume %s not found, assuming it to be already deleted", volID)
-			return nil
-		}
-	}
-
-	if err := os.RemoveAll(volRootDeleting); err != nil {
-		return fmt.Errorf("failed to delete volume %s: %v", volID, err)
-	}
-
-	return nil
-}
-
-func mountCephRoot(volID volumeID, volOptions *volumeOptions, adminCr *credentials) error {
-	cephRoot := getCephRootPathLocal(volID)
-
-	// Root path is not set for dynamically provisioned volumes
-	// Access to cephfs's / is required
-	volOptions.RootPath = "/"
-
-	if err := createMountPoint(cephRoot); err != nil {
-		return err
-	}
-
-	m, err := newMounter(volOptions)
-	if err != nil {
-		return fmt.Errorf("failed to create mounter: %v", err)
-	}
-
-	if err = m.mount(cephRoot, adminCr, volOptions); err != nil {
-		return fmt.Errorf("error mounting ceph root: %v", err)
-	}
+func createVolume(volOptions *volumeOptions, volID volumeID, bytesQuota int64) error {
+	_, _, err := util.ExecCommand(
+		"ceph",
+		"fs",
+		"subvolume",
+		"create",
+		volOptions.FsName,
+		string(volID),
+		string(bytesQuota))
 
-	return nil
+	return err
 }
 
-func unmountCephRoot(volID volumeID) {
-	cephRoot := getCephRootPathLocal(volID)
+func purgeVolume(volID volumeID, volOptions *volumeOptions) error {
+	_, _, err := util.ExecCommand(
+		"ceph",
+		"fs",
+		"subvolume",
+		"rm",
+		volOptions.FsName,
+		string(volID))
 
```

Comment:
How do you handle the case when trying to purge_subvolume that is already removed as was done earlier?  This command will raise ENOENT if the subvolume doesn't exist. To suppress ENOENT error, you can add do,
`ceph fs subvolume rm --force`